Python 3.15t segfault when importing numpy.random

The 3.15t pathfinder "see what works" test segfaults
```
Running pathfinder tests with  LD:see_what_works  FH:see_what_works  BC:see_what_works
/__w/cuda-python/cuda-python/ci/tools/run-tests: line 155:  4249 Segmentation fault      (core dumped) pytest -ra -s -v --durations=0 tests/ 2>&1
      4250 Done                    | tee /tmp/pathfinder_test_log.txt
Error: Process completed with exit code 139.
```
https://github.com/NVIDIA/cuda-python/actions/runs/27041829125/job/79822100279?pr=2108#step:28:75

_Originally posted by @leofang in https://github.com/NVIDIA/cuda-python/issues/2108#issuecomment-4636010325_

# Issue 2171 report: Python 3.15t segfault reproduced locally

## Human-targeted summary

I was able to reproduce the issue 2171 segfault locally by using the exact GitHub Actions Python distribution from `actions/python-versions`, rather than a locally built CPython 3.15t.

The failure does not currently look like a `cuda_pathfinder` implementation bug. The reduced reproducer is:

```bash
python -X faulthandler -c 'from numpy import random'
```

That segfaults with:

- GitHub Actions Python: `Python 3.15.0b2 free-threading build (main, Jun 3 2026, 14:43:48) [GCC 11.4.0]`
- NumPy wheel: `numpy-2.4.6-cp315-cp315t-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl`
- GIL disabled / free-threaded Python

The `cuda_pathfinder` test crash is an indirect trigger:

1. `pytest` starts.
2. The `pytest-randomly` plugin is auto-loaded.
3. `pytest-randomly` imports `from numpy import random`.
4. Importing `numpy.random` segfaults before pytest prints its session header.

As a practical workaround, disabling the `pytest-randomly` plugin avoids the crash:

```bash
python -m pytest -p no:randomly -ra -s -v --durations=0 tests/
```

With that workaround, the full local `cuda_pathfinder` test suite passed under the same reproduced environment:

```text
973 passed, 6 skipped in 6.71s
```

This suggests issue 2171 should probably be tracked as an incompatibility between the GitHub Actions Python 3.15t beta build and the custom NumPy 2.4.6 cp315t wheel, exposed by `pytest-randomly`, rather than as a CUDA pathfinder library discovery failure.

## Agent-facing full details

### Original failing CI job

The original CI failure was in the PR 2108 GitHub Actions job:

```text
https://github.com/NVIDIA/cuda-python/actions/runs/27041829125/job/79822100279?pr=2108
```

The manually downloaded CI job log is available locally at:

```text
/wrk/obox/cuda-python_pr2108_run27041829125_job79822100279_linux64_py315t_cuda1330_pathfinder_segfault.log
```

The key CI failure was:

```text
/__w/cuda-python/cuda-python/ci/tools/run-tests: line 155: 4249 Segmentation fault (core dumped) pytest -ra -s -v --durations=0 tests/ 2>&1
##[error]Process completed with exit code 139.
```

The CI environment installed only `cuda_pathfinder`'s basic `test` dependency group before the crash, not `test-cu13`. It also installed a custom NumPy prerelease/free-threaded wheel:

```text
numpy-2.4.6-cp315-cp315t-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
```

### Local reproduction setup

At first, local tests passed with a locally built CPython 3.15t:

```text
Python 3.15.0b2 free-threading build (tags/v3.15.0b2:94a64bbc6ce, Jun 8 2026, 20:42:06) [GCC 13.3.0]
```

That local build did not reproduce the segfault, even after matching:

- the basic `test` dependency group
- the custom NumPy 2.4.6 cp315t wheel
- the `pytest` command line
- the pathfinder test strictness environment variables
- `LD_LIBRARY_PATH` pointing to the Python `lib` directory
- a local setup where the system CUDA toolkit was hidden via `local_cuda.sh off`

The decisive difference was the Python distribution itself.

### Installing the exact GitHub Actions Python

The exact GitHub Actions Python asset was downloaded from:

```text
actions/python-versions release: 3.15.0-beta.2-26891732290
asset: python-3.15.0-beta.2-linux-22.04-x64-freethreaded.tar.gz
```

The successful install log is:

```text
/wrk/logs/install_actions-toolcache_python-3.15.0-beta.2-linux-22.04-x64-freethreaded_log_2026-06-09+094635.txt
```

Successful install summary:

```text
Create Python 3.15.0-beta.2 folder
Copy Python binaries to hostedtoolcache folder
Create additional symlinks
Upgrading pip...
Successfully installed pip-26.1.2
Create complete file
# EXIT status=0
```

Installed local toolcache path:

```text
/wrk/cpython_installs/actions-toolcache/Python/3.15.0-beta.2/x64-freethreaded
```

The local scripts were then adjusted to use that Python:

```text
/wrk/build_pathfinder_py315t.sh
/wrk/test_pathfinder_py315t.sh
```

The reproduced build log is:

```text
/wrk/logs/build_pathfinder_py315t_log_2026-06-09+095227.txt
```

Important build details:

```text
Python 3.15.0b2 free-threading build (main, Jun 3 2026, 14:43:48) [GCC 11.4.0]
Successfully installed numpy-2.4.6
Successfully installed cuda-pathfinder-1.5.6.dev19+g4bb52eb13
Successfully installed pytest-9.0.3 pytest-mock-3.15.1 pytest-randomly-4.1.0 pytest-repeat-0.9.4
```

Installed packages in the reproduced environment:

```text
cuda-pathfinder 1.5.6.dev19+g4bb52eb13
iniconfig       2.3.0
numpy           2.4.6
packaging       26.2
pip             26.1.2
pluggy          1.6.0
Pygments        2.20.0
pytest          9.0.3
pytest-mock     3.15.1
pytest-randomly 4.1.0
pytest-repeat   0.9.4
```

### Local segfault reproduction

The reproduced failing test log is:

```text
/wrk/logs/test_pathfinder_py315t_log_2026-06-09+095255.txt
```

The failure is very early:

```text
+ pytest -ra -s -v --durations=0 tests/
/wrk/test_pathfinder_py315t.sh: line 15: 471521 Segmentation fault (core dumped) pytest -ra -s -v --durations=0 tests/
# EXIT status=139
```

There is no pytest session header. This means the crash happens during Python / pytest / plugin startup, before pytest begins normal collection output.

### Reduced reproducer

The first minimal reduction showed:

```bash
python -VV
# ok

python -c 'import sys; print(sys._is_gil_enabled())'
# False

python -c 'import pytest; print(pytest.__version__)'
# ok, 9.0.3

python -c 'import numpy; print(numpy.__version__)'
# ok, 2.4.6

python -c 'import pytest_randomly; print(pytest_randomly.__version__)'
# segfault
```

`pytest-randomly` imports NumPy random at module import time:

```text
/wrk/forked/cuda-python/Py3150b2Venv/lib/python3.15t/site-packages/pytest_randomly/__init__.py
```

Relevant code:

```python
# numpy
try:
    from numpy import random as np_random

    have_numpy = True
except ImportError:  # pragma: no cover
    have_numpy = False
```

Further reduction confirms `pytest-randomly` is not necessary for the crash:

```bash
python -X faulthandler -c 'from numpy import random'
python -X faulthandler -c 'import numpy.random'
python -X faulthandler -c 'import numpy.random._pickle'
python -X faulthandler -c 'import numpy.random.mtrand'
```

These all segfaulted in the reproduced environment.

### Faulthandler evidence

Representative command:

```bash
python -X faulthandler -c 'import pytest_randomly'
```

Representative Python stack:

```text
Fatal Python error: Segmentation fault

<Cannot show all threads while the GIL is disabled>
Stack (most recent call first):
  File "/wrk/forked/cuda-python/Py3150b2Venv/lib/python3.15t/site-packages/numpy/random/_pickle.py", line 7 in <module>
  File "/wrk/forked/cuda-python/Py3150b2Venv/lib/python3.15t/site-packages/numpy/random/__init__.py", line 180 in <module>
  File "/wrk/forked/cuda-python/Py3150b2Venv/lib/python3.15t/site-packages/numpy/__init__.py", line 718 in __getattr__
  File "/wrk/forked/cuda-python/Py3150b2Venv/lib/python3.15t/site-packages/pytest_randomly/__init__.py", line 51 in <module>
  File "<string>", line 1 in <module>
```

Representative native stack components:

```text
libpython3.15t.so.1.0
numpy/_core/_multiarray_umath.cpython-315t-x86_64-linux-gnu.so
numpy/random/bit_generator.cpython-315t-x86_64-linux-gnu.so
numpy/random/_mt19937.cpython-315t-x86_64-linux-gnu.so
numpy/random/mtrand.cpython-315t-x86_64-linux-gnu.so
```

Loaded extension modules at crash time:

```text
numpy._core._multiarray_umath
numpy.linalg._umath_linalg
numpy.random._common
numpy.random.bit_generator
numpy.random._bounded_integers
numpy.random._pcg64
numpy.random._generator
numpy.random._mt19937
numpy.random._philox
numpy.random._sfc64
numpy.random.mtrand
```

### Workaround verification

Disabling the `pytest-randomly` plugin avoids importing `numpy.random` during pytest startup:

```bash
python -m pytest -p no:randomly -ra -s -v --durations=0 tests/
```

This passed locally in the reproduced environment:

```text
973 passed, 6 skipped in 6.71s
```

The skipped tests were expected in this local environment:

```text
SKIPPED [3] tests/conftest.py:43: libnvcudla.so is not loadable via canary subprocess on this host.
SKIPPED [2] tests/test_find_nvidia_headers.py:127: nvidia-cutlass not available on Python 3.15 (scipy missing)
SKIPPED [1] tests/test_utils_env_vars.py:93: Windows-specific case-folding check
```

### Current interpretation

The most likely causal chain is:

```text
GitHub Actions Python 3.15t beta build
+ custom NumPy 2.4.6 cp315t wheel
+ import numpy.random
= native segfault
```

The `cuda_pathfinder` test suite is not required to reproduce the crash. It only encounters the issue because its `test` dependency group includes `pytest-randomly`, and pytest auto-loads installed plugins by default.

The reason local testing originally passed is that the locally built CPython 3.15t was not identical to the GitHub Actions Python distribution:

```text
Local non-crashing Python:
Python 3.15.0b2 free-threading build (tags/v3.15.0b2:94a64bbc6ce, Jun 8 2026, 20:42:06) [GCC 13.3.0]

Actions crashing Python:
Python 3.15.0b2 free-threading build (main, Jun 3 2026, 14:43:48) [GCC 11.4.0]
```

### Suggested next steps

For unblocking cuda-python CI, the narrow workaround is to disable `pytest-randomly` for Python 3.15t jobs, for example by passing:

```bash
-p no:randomly
```

or by setting:

```bash
PYTEST_DISABLE_PLUGIN_AUTOLOAD=1
```

and explicitly loading only required plugins. The `-p no:randomly` workaround is smaller and was verified locally.

For root cause, report the minimal reproducer to the owner of the custom NumPy cp315t wheel / NumPy prerelease build process:

```bash
python -X faulthandler -c 'from numpy import random'
```

with these details:

```text
Python: GitHub Actions python-3.15.0-beta.2-linux-22.04-x64-freethreaded
Python version string: Python 3.15.0b2 free-threading build (main, Jun 3 2026, 14:43:48) [GCC 11.4.0]
GIL enabled: False
NumPy: 2.4.6 cp315t custom wheel
Crash module path: numpy.random / numpy.random.mtrand / numpy.random.bit_generator
```

It may also be useful to compare against:

- a newer `actions/python-versions` 3.15t build, if available
- a locally rebuilt NumPy wheel against the exact Actions Python
- the same wheel under the locally built CPython 3.15t that did not crash
- Python 3.15 non-free-threaded, if a comparable NumPy wheel is available

--- a/.github/workflows/ci-pixi-source-test.yml
+++ b/.github/workflows/ci-pixi-source-test.yml
@@ -112,7 +112,7 @@ jobs:
     timeout-minutes: 90
     container:
       options: -u root --security-opt seccomp=unconfined --shm-size 16g
-      image: ubuntu:22.04
+      image: ubuntu:24.04
     steps:
       - name: Ensure GPU is working
         run: nvidia-smi

--- a/.github/workflows/coverage.yml
+++ b/.github/workflows/coverage.yml
@@ -47,7 +47,7 @@ jobs:
     # TODO: use a different (nvidia?) container
     container:
       options: -u root --security-opt seccomp=unconfined --shm-size 16g
-      image: ubuntu:22.04
+      image: ubuntu:24.04
       env:
         NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
         PIP_CACHE_DIR: "/tmp/pip-cache"

--- a/.github/workflows/test-wheel-linux.yml
+++ b/.github/workflows/test-wheel-linux.yml
@@ -118,7 +118,7 @@ jobs:
       # can nsenter to the host for the install + refresh the toolkit bind mounts
       # back inside the container. Stock options for latest/earliest rows.
       options: ${{ ((matrix.DRIVER == 'latest' || matrix.DRIVER == 'earliest') && '-u root --security-opt seccomp=unconfined --shm-size 16g') || '-u root --security-opt seccomp=unconfined --shm-size 16g --privileged --pid=host' }}
-      image: ubuntu:22.04
+      image: ubuntu:24.04
       env:
         NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
         PIP_CACHE_DIR: "/tmp/pip-cache"

--- a/ci/test-matrix.yml
+++ b/ci/test-matrix.yml
@@ -52,8 +52,7 @@ linux:
     - { ARCH: 'amd64', PY_VER: '3.14t', CUDA_VER: '13.0.2', LOCAL_CTK: '1', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }
     - { ARCH: 'amd64', PY_VER: '3.14t', CUDA_VER: '13.3.0', LOCAL_CTK: '1', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }
     - { ARCH: 'amd64', PY_VER: '3.15',  CUDA_VER: '13.3.0', LOCAL_CTK: '0', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }
-    # Blocked by NVIDIA/cuda-python#2171
-    # - { ARCH: 'amd64', PY_VER: '3.15t', CUDA_VER: '13.3.0', LOCAL_CTK: '0', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }
+    - { ARCH: 'amd64', PY_VER: '3.15t', CUDA_VER: '13.3.0', LOCAL_CTK: '0', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }
     # linux-aarch64
     - { ARCH: 'arm64', PY_VER: '3.10',  CUDA_VER: '12.9.1', LOCAL_CTK: '1', GPU: 'a100',       GPU_COUNT: '1', DRIVER: 'latest' }
     - { ARCH: 'arm64', PY_VER: '3.10',  CUDA_VER: '13.0.2', LOCAL_CTK: '0', GPU: 'l4',         GPU_COUNT: '1', DRIVER: 'latest' }

